File: python/pydroid_flow/engine_parts/nodes/conversion_ui.py
from __future__ import annotations

import io
import json
from typing import Any

import numpy as np
import pandas as pd

from ..custom_function import _execute_custom_function
from ..presentation import _printable
from ..values import _as_bool, _round_half_away, _single_value, _require_table

NODE_TYPES = {
    "io.export_csv",
    "convert.to_text",
    "convert.to_number",
    "convert.to_boolean",
    "convert.to_table",
    "convert.table_to_records",
    "convert.table_to_csv",
    "convert.json_parse",
    "convert.json_stringify",
    "python.len",
    "python.round",
    "python.print",
    "ui.alert",
    "ui.input_dialog",
    "custom.python_function",
}


def execute(
    node_type: str,
    params: dict[str, Any],
    upstream: Any,
    csv_text: str,
    input_files: list[dict[str, Any]],
    variables: dict[str, Any] | None = None,
) -> tuple[dict[str, Any], pd.DataFrame | None, str | None, str | None]:
    del csv_text, input_files, variables
    table_result: pd.DataFrame | None = None
    export_result: str | None = None

    if node_type == "io.export_csv":
        export_result = _require_table(upstream, "Export CSV").to_csv(index=False, lineterminator="\n")
        value = export_result
    elif node_type == "convert.to_text":
        value = _printable(upstream) if _as_bool(params.get("pretty", True)) else str(upstream)
    elif node_type == "convert.to_number":
        raw = _single_value(upstream)
        number = float(raw)
        value = _round_half_away(number) if _as_bool(params.get("integer", False)) else number
    elif node_type == "convert.to_boolean":
        raw = _single_value(upstream)
        if isinstance(raw, str):
            token = raw.strip().lower()
            if token in {"true", "1", "yes", "y", "是", "真"}:
                value = True
            elif token in {"false", "0", "no", "n", "否", "假", "", "none", "null"}:
                value = False
            else:
                raise ValueError(f"无法将文本 {raw!r} 转换为布尔值")
        else:
            value = bool(raw)
    elif node_type == "convert.to_table":
        if isinstance(upstream, pd.DataFrame):
            value = upstream.copy()
        elif isinstance(upstream, pd.Series):
            value = upstream.to_frame()
        elif _as_bool(params.get("csvText", False)) and isinstance(upstream, str):
            value = pd.read_csv(io.StringIO(upstream))
        elif isinstance(upstream, dict):
            try:
                value = pd.DataFrame(upstream)
            except ValueError:
                value = pd.DataFrame([upstream])
        elif isinstance(upstream, (list, tuple, np.ndarray)):
            value = pd.DataFrame(upstream)
        else:
            value = pd.DataFrame({"value": [upstream]})
        table_result = value
    elif node_type == "convert.table_to_records":
        value = _require_table(upstream, "Table to records").to_dict(orient="records")
    elif node_type == "convert.table_to_csv":
        value = _require_table(upstream, "Table to CSV").to_csv(index=_as_bool(params.get("includeIndex", False)), lineterminator="\n")
    elif node_type == "convert.json_parse":
        value = json.loads(str(upstream))
    elif node_type == "convert.json_stringify":
        value = json.dumps(upstream, ensure_ascii=False, indent=max(0, min(8, int(params.get("indent", 2)))), default=str)
    elif node_type == "python.len":
        value = len(upstream)
    elif node_type == "python.round":
        if not isinstance(upstream, (int, float)):
            raise ValueError("Python round requires a numeric input")
        value = _round_half_away(upstream, int(params.get("digits", 0)))
    elif node_type == "python.print":
        prefix = str(params.get("prefix", "")).strip()
        rendered = _printable(upstream, max(100, int(params.get("maxChars", 8000))), max(1, int(params.get("maxRows", 20))), str(params.get("format", "pretty")), _as_bool(params.get("includeType", True)), str(params.get("encoding", "utf-8")), str(params.get("encodingErrors", "replace")), str(params.get("bytesFormat", "decode")))
        rendered = (f"{prefix}：" if prefix else "") + rendered + str(params.get("end", ""))
        return {"output": upstream, "__print__": rendered}, table_result, None, export_result
    elif node_type == "ui.alert":
        content = upstream.get("content") if isinstance(upstream, dict) else upstream
        rendered = f"{str(params.get('title', '提示')).strip()}：{str(params.get('message', '')).strip()}"
        if content is not None:
            rendered += "\n" + _printable(content, 4000, 20, "pretty", True)
        response = params.get("response")
        reported = f"{rendered}\n选择：{response!r}"
        return {"output": response, "__print__": reported[:1000]}, table_result, None, export_result
    elif node_type == "ui.input_dialog":
        raw_value = params.get("value", "")
        input_kind = str(params.get("inputKind", "text"))
        if input_kind == "number":
            try:
                value = float(raw_value)
                if value.is_integer():
                    value = int(value)
            except (TypeError, ValueError):
                raise ValueError("弹窗输入节点需要有效数值")
        elif input_kind == "boolean":
            value = _as_bool(raw_value)
        elif input_kind == "json":
            try:
                value = json.loads(str(raw_value))
            except json.JSONDecodeError as exception:
                raise ValueError(f"弹窗输入的 JSON 无效：{exception.msg}") from exception
        elif input_kind == "table":
            text = str(raw_value).strip()
            try:
                value = pd.DataFrame(json.loads(text))
            except (json.JSONDecodeError, TypeError, ValueError):
                value = pd.read_csv(io.StringIO(text), sep=None, engine="python")
            table_result = value
        else:
            value = str(raw_value)
    elif node_type == "custom.python_function":
        outputs = _execute_custom_function(str(params.get("code", "")), upstream, params)
        table_result = next((item for item in outputs.values() if isinstance(item, pd.DataFrame)), None)
        return outputs, table_result, None, export_result
    else:
        raise ValueError(f"Unsupported conversion/UI node type: {node_type}")

    return {"output": value}, table_result, None, export_result

File: src/runtime/javascript/engine/nodes/conversion_ui.ts
import { parseCsv } from "../csv";
import { executeCustomFunction, executeJsCell } from "../notebook";
import { printable, singleValue } from "../printable";
import { Table, tableFromValue } from "../table";
import { asBool, requireTable } from "./support/common";
import { jsonSafe } from "./support/serialization";
import type { ExecutionContext, NodeOutput } from "./support/types";

export function executeConversionUiNode(nodeType: string, params: Record<string, unknown>, upstream: unknown, context: ExecutionContext): NodeOutput | null {
  const tableResult: Table | null = null;
  const plotResult = null;
  const exportResult: string | null = null;
  const table = (): Table => requireTable(upstream, "Table input");

  switch (nodeType) {
    case "io.export_csv": {
      const content = `${table().toCSV(false, "\n")}\n`;
      return { outputs: { output: content }, tableResult, plotResult, exportResult: content };
    }
    case "convert.to_text": {
      const value = asBool(params.pretty ?? true) ? printable(upstream) : String(upstream);
      return { outputs: { output: value }, tableResult, plotResult, exportResult };
    }
    case "convert.to_number": {
      const raw = singleValue(upstream);
      const value = asBool(params.integer ?? false) ? Math.trunc(Number(raw)) : Number(raw);
      return { outputs: { output: value }, tableResult, plotResult, exportResult };
    }
    case "convert.to_boolean": {
      const raw = singleValue(upstream);
      let value: boolean;
      if (typeof raw === "string") {
        const token = raw.trim().toLowerCase();
        if (["true", "1", "yes", "y", "是", "真"].includes(token)) value = true;
        else if (["false", "0", "no", "n", "否", "假", "", "none", "null"].includes(token)) value = false;
        else throw new Error(`无法将文本 ${raw} 转换为布尔值`);
      } else {
        value = Boolean(raw);
      }
      return { outputs: { output: value }, tableResult, plotResult, exportResult };
    }
    case "convert.to_table": {
      let value: Table;
      if (upstream instanceof Table) value = upstream.copy();
      else if (asBool(params.csvText ?? false) && typeof upstream === "string") value = parseCsv(upstream);
      else if (upstream && typeof upstream === "object" && !Array.isArray(upstream)) {
        try {
          value = Table.fromObject(upstream as Record<string, unknown>);
        } catch {
          value = new Table(["value"], [[JSON.stringify(upstream)]]);
        }
      }
      else if (Array.isArray(upstream)) value = tableFromValue(upstream, "Table");
      else value = new Table(["value"], [[upstream ?? null]]);
      return { outputs: { output: value }, tableResult: value, plotResult, exportResult };
    }
    case "convert.table_to_records": {
      const value = table().records();
      return { outputs: { output: value }, tableResult, plotResult, exportResult };
    }
    case "convert.table_to_csv": {
      const value = `${table().toCSV(asBool(params.includeIndex ?? false), "\n")}\n`;
      return { outputs: { output: value }, tableResult, plotResult, exportResult };
    }
    case "convert.json_parse": {
      return { outputs: { output: JSON.parse(String(upstream)) }, tableResult, plotResult, exportResult };
    }
    case "convert.json_stringify": {
      const indent = Math.max(0, Math.min(8, Number(params.indent ?? 2)));
      const safe = jsonSafe(upstream);
      // Python json.dumps(..., indent=0) is newline-formatted with zero leading
      // indentation; JSON.stringify(..., null, 0) is compact. Preserve Python
      // semantics so the node is runtime-neutral.
      const value = indent === 0
        ? JSON.stringify(safe, null, 1).split("\n").map((line) => line.trimStart()).join("\n")
        : JSON.stringify(safe, null, indent);
      return { outputs: { output: value }, tableResult, plotResult, exportResult };
    }
    case "python.len": {
      if (upstream instanceof Table) return { outputs: { output: upstream.rowCount }, tableResult, plotResult, exportResult };
      if (Array.isArray(upstream) || typeof upstream === "string") return { outputs: { output: upstream.length }, tableResult, plotResult, exportResult };
      if (upstream && typeof upstream === "object") return { outputs: { output: Object.keys(upstream).length }, tableResult, plotResult, exportResult };
      throw new Error("python.len requires a sized input");
    }
    case "python.round": {
      if (typeof upstream !== "number") throw new Error("Python round requires a numeric input");
      const digits = Number(params.digits ?? 0);
      const factor = 10 ** digits;
      return { outputs: { output: Math.round(upstream * factor) / factor }, tableResult, plotResult, exportResult };
    }
    case "python.print": {
      const prefix = String(params.prefix ?? "").trim();
      const rendered = printable(upstream, Number(params.maxChars ?? 8000), Number(params.maxRows ?? 20), String(params.format ?? "pretty") as "pretty" | "repr" | "text" | "json", asBool(params.includeType ?? true));
      const text = (prefix ? `${prefix}：` : "") + rendered + String(params.end ?? "");
      return { outputs: { output: upstream, __print__: text }, tableResult, plotResult, exportResult };
    }
    case "ui.alert": {
      const content = upstream && typeof upstream === "object" && !Array.isArray(upstream) ? (upstream as Record<string, unknown>).content : upstream;
      const rendered = `${String(params.title ?? "提示").trim()}：${String(params.message ?? "").trim()}` + (content !== null && content !== undefined ? `\n${printable(content, 4000, 20)}` : "");
      const response = params.response;
      const reported = `${rendered}\n选择：${String(response)}`;
      return { outputs: { output: response, __print__: reported.slice(0, 1000) }, tableResult, plotResult, exportResult };
    }
    case "ui.input_dialog": {
      const rawValue = params.value ?? "";
      const inputKind = String(params.inputKind ?? "text");
      let value: unknown;
      let tableResultValue: Table | null = null;
      if (inputKind === "number") {
        const number = Number(rawValue);
        if (Number.isNaN(number)) throw new Error("弹窗输入节点需要有效数值");
        value = Number.isInteger(number) ? number : number;
      } else if (inputKind === "boolean") {
        value = asBool(rawValue);
      } else if (inputKind === "json") {
        try {
          value = JSON.parse(String(rawValue));
        } catch (error) {
          throw new Error(`弹窗输入的 JSON 无效：${(error as Error).message}`);
        }
      } else if (inputKind === "table") {
        const text = String(rawValue).trim();
        try {
          value = Table.fromRecords(JSON.parse(text) as Array<Record<string, unknown>>);
        } catch {
          value = parseCsv(text, { header: 0 });
        }
        tableResultValue = value as Table;
      } else {
        value = String(rawValue);
      }
      return { outputs: { output: value }, tableResult: tableResultValue, plotResult, exportResult };
    }
    case "custom.python_function": {
      const outputs = executeCustomFunction(String(params.code ?? ""), (upstream as Record<string, unknown>) ?? {}, params);
      const tableValue = Object.values(outputs).find((item) => item instanceof Table);
      return { outputs, tableResult: tableValue instanceof Table ? tableValue : null, plotResult, exportResult };
    }
    case "notebook.code_cell": {
      const source = String(params.source ?? "");
      const result = executeJsCell(source, context.notebookNamespace);
      return { outputs: result.outputs, tableResult: result.table, plotResult: result.plot, exportResult };
    }
    case "notebook.markdown_cell": {
      const text = String(params.source ?? "");
      return { outputs: { next: text, output: text }, tableResult, plotResult, exportResult };
    }
    default:
      return null;
  }
}
